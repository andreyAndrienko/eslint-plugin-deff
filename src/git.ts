import * as child_process from "child_process";
import * as path from "path";
import { Range } from "./Range";

const sanitizeFilePath = (filePath: string) =>
  JSON.stringify(path.resolve(filePath));

const diffCacheKey = (filePath: string, staged: boolean): string =>
  JSON.stringify([path.resolve(filePath), staged]);

const setCachedDiff = (filePath: string, staged: boolean, diff: string): void =>
  void diffCache.set(diffCacheKey(filePath, staged), diff);

const getCachedDiff = (filePath: string, staged: boolean) =>
  diffCache.get(diffCacheKey(filePath, staged));

const diffCache = new Map<string, string>();
const getDiffForFile = (filePath: string, staged = false): string => {
  let diff = getCachedDiff(filePath, staged);
  if (diff === undefined) {
    const command = [
      "git",
      "diff",
      "--diff-filter=ACM",
      staged && "--staged",
      "--unified=0",
      JSON.stringify(process.env.ESLINT_PLUGIN_DIFF_COMMIT ?? "HEAD"),
      "--",
      sanitizeFilePath(filePath),
    ]
      .filter(Boolean)
      .join(" ");

    const result = child_process.execSync(command).toString();
    setCachedDiff(filePath, staged, result);
    diff = result;
  }

  return diff;
};

let diffFileListCache: string[] | undefined;
const getDiffFileList = (staged = false): string[] => {
  if (diffFileListCache === undefined) {
    const command = [
      "git",
      "diff",
      "--diff-filter=ACM",
      "--name-only",
      staged && "--staged",
      JSON.stringify(process.env.ESLINT_PLUGIN_DIFF_COMMIT ?? "HEAD"),
    ]
      .filter(Boolean)
      .join(" ");

    diffFileListCache = child_process
      .execSync(command)
      .toString()
      .trim()
      .split("\n")
      .map((filePath) => path.resolve(filePath));
  }
  return diffFileListCache;
};

let gitFileListCache: string[] | undefined;
const getGitFileList = (): string[] => {
  if (gitFileListCache === undefined) {
    const command = ["git", "ls-files"].filter(Boolean).join(" ");

    gitFileListCache = child_process
      .execSync(command)
      .toString()
      .trim()
      .split("\n")
      .map((filePath) => path.resolve(filePath));
  }
  return gitFileListCache;
};

const getIgnorePatterns = (staged = false): string[] => {
  const changedFiles = getDiffFileList(staged);

  const unchangedFiles = getGitFileList()
    .filter((x) => !changedFiles.includes(x))
    .map((x) => path.join("/", path.relative(process.cwd(), x)));

  return unchangedFiles;
};

const isHunkHeader = (input: string) => {
  const hunkHeaderRE = new RegExp(/^@@ .* @@/g);
  return input.match(hunkHeaderRE);
};

const getRangeForChangedLines = (line: string) => {
  const rangeRE = new RegExp(
    /^@@ .* \+(?<start>\d+)(?<linesCountDelimiter>,(?<linesCount>\d+))? @@/
  );
  const range = rangeRE.exec(line);
  if (range === null) {
    throw Error(`Couldn't match regex with line '${line}'`);
  }
  if (range.groups?.start === undefined) {
    /*
     * NOTE: Never happens, because RegExp requires start to be a
     * required number
     */
    throw Error("Couldn't match regex to find start");
  }

  const linesCount: number =
    range.groups.linesCountDelimiter && range.groups.linesCount
      ? parseInt(range.groups.linesCount)
      : 1;

  const hasAddedLines = linesCount !== 0;
  const start: number = parseInt(range.groups.start);
  const end = start + linesCount;

  return hasAddedLines ? new Range(start, end) : null;
};

const removeNullRanges = (r: Range | null): r is Range => r !== null;

const getRangesForDiff = (diff: string): Range[] =>
  diff
    .split("\n")
    .filter(isHunkHeader)
    .map(getRangeForChangedLines)
    .filter(removeNullRanges);

export { getDiffForFile, getIgnorePatterns, getRangesForDiff, getDiffFileList };
export type { Range };
