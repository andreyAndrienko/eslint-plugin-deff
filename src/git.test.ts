import * as child_process from "child_process";
import { mocked } from "ts-jest/utils";
import { getDiffFileList, getDiffForFile, getRangesForDiff } from "./git";
import {
  hunks,
  includingOnlyRemovals,
  diffFileList,
} from "./__fixtures__/diff";
import path from "path";

jest.mock("child_process");

const mockedChildProcess = mocked(child_process, true);

describe("git", () => {
  it("should find the ranges of each staged file", () => {
    expect(getRangesForDiff(hunks)).toMatchSnapshot();
  });

  it("should work for hunks which include only-removal-ranges", () => {
    expect(getRangesForDiff(includingOnlyRemovals)).toMatchSnapshot();
  });

  it("should get the staged diff of a file", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(hunks));

    const diffFromFile = getDiffForFile("./mockfile.js");

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
    expect(diffFromFile).toContain("diff --git");
    expect(diffFromFile).toContain("@@");
  });

  it("should hit the cached diff of a file", () => {
    jest.mock("child_process").resetAllMocks();
    mockedChildProcess.execSync.mockReturnValueOnce(Buffer.from(hunks));

    const diffFromFileA = getDiffForFile("./mockfileCache.js");
    const diffFromFileB = getDiffForFile("./mockfileCache.js");
    expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(1);
    expect(diffFromFileA).toEqual(diffFromFileB);

    mockedChildProcess.execSync.mockReturnValueOnce(Buffer.from(hunks));
    getDiffForFile("./mockfileMiss.js");
    expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(2);
  });

  it("should get the list of staged files", () => {
    jest.mock("child_process").resetAllMocks();
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(diffFileList));

    const fileList = getDiffFileList();

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
    expect(fileList).toEqual(
      ["file1", "file2", "file3"].map((p) => path.resolve(p))
    );
  });
});
