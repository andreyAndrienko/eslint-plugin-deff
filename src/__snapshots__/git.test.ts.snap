// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git should find the ranges of each staged file 1`] = `
Array [
  Range {
    "exclusiveUpperBound": 1,
    "inclusiveLowerBound": 0,
  },
  Range {
    "exclusiveUpperBound": 6,
    "inclusiveLowerBound": 2,
  },
  Range {
    "exclusiveUpperBound": 9,
    "inclusiveLowerBound": 7,
  },
]
`;

exports[`git should work for hunks which include only-removal-ranges 1`] = `
Array [
  Range {
    "exclusiveUpperBound": 4,
    "inclusiveLowerBound": 2,
  },
]
`;
